src/adminUtils.js

```diff
@@ -0,0 +1,199 @@
+  }
+
+  _admins() {
+    return nconf.get('adminList');
+  }
+
+  _get_voice_channel(message) {
+    if (message.member && message.member.voiceChannel) {
+      return message.member.voiceChannel;
+    }
+
+    message.reply("You don't appear to be in a voice channel!");
+  }
+  _get_discord_user(message, username) {
+    return message.channel.guild.members.find(a => {
+      return a.user['username'].toLowerCase() == username.toLowerCase();
+    });
+  }
+
+  _saveConfig(key, value) {
+    log.debug(`Saving config key: ${key}, value: ${value}`);
+    nconf.set(key, value);
+    const configFile = nconf.stores.file.file;
+    nconf.save(err => {
+      fs.readFile(configFile, (err, data) => {
+        data = data || {};
+        // TODO: I'm pretty sure this can be dropped
+        console.dir(JSON.parse(data.toString()));
+      })
+    });
+  }
+
+  // Public functions
+  access(message, params) {
+    if (!params.length > 0) {
+      message.reply("Not enough details to show access");
+      return;
+    }
+
+    const username = params[0];
+    const discord_user = this._get_discord_user(message, username);
+    if (discord_user && this._admins()[discord_user.user.id]) {
+      this._print_access(message, username, discord_user.user.id);
+    } else {
+      message.reply(`${username} does not presently have any admin permissions`)
+    }
+  }
+
+  add(message, params) {
+    if (!params.length > 0) {
+      return;
+    }
+
+    const prefix = params[0];
+    if (!prefix.match(/^[a-z0-9_]+$/)) {
+      message.reply(`${prefix} is a bad short name`);
+    }
+
+    if (message.attachments.first()) {
+      // Only check the first attachment
+      const a = message.attachments.first();
+      const filename = `./Uploads/${prefix}--${a.filename}`;
+
+      log.debug(`Writing attachment to file: ${filename}`);
+      request(a.url).pipe(fs.createWriteStream(filename));
+
+      files[prefix] = filename;
+      message.reply(`!${prefix} is now available`);
+    } else {
+      message.reply(`You need to attach a file`);
+    }
+  }
+
+  grant(message, params) {
+    if (!params.length > 0) {
+      message.reply("Not enough details to add access");
+      return;
+    }
+
+    const username = params[0];
+    let access = params[1];
+    const discord_user = this._get_discord_user(message, username);
+    const adminList = this._admins();
+    if (discord_user && access) {
+      log.debug(`Updating: ${username} with ${access}`);
+      access = access.split(',').map(operation => operation.trim());
+      console.log(access);
+      const userId = discord_user.user.id;
+
+      if (!(userId in adminList)) {
+        log.debug("New user")
+        adminList[userId] = {'access': access, 'immune': false};
+      } else {
+        log.debug("Additional permissions")
+        adminList[userId]['access'] = [...access, ...adminList[userId]['access']]
+          .filter((element, pos, arr) => {
```

Comment:
Given the join code above, I don't think this should be necessary - it already dedupes.

Reply:
I found a situation in which users could actually end up with multiple 'adds' etc..., so this was necessary. I think it's because if you already have add we still run this logic, so it has the new 'add' and the existing 'add'.